sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReceiveLinkProcessor.java

```diff
@@ -14,47 +14,69 @@
 import reactor.core.CoreSubscriber;
 import reactor.core.Disposable;
 import reactor.core.Disposables;
+import reactor.core.Exceptions;
 import reactor.core.publisher.FluxProcessor;
 import reactor.core.publisher.Mono;
 import reactor.core.publisher.Operators;
 import reactor.core.scheduler.Schedulers;
 
 import java.time.Duration;
+import java.util.Deque;
 import java.util.Objects;
+import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 
 import static com.azure.core.util.FluxUtil.monoError;
 
 /**
  * Processes AMQP receive links into a stream of AMQP messages.
+ *
+ * This is almost a carbon copy of AmqpReceiveLinkProcessor. When we can abstract it from proton-j, it would be nice to
+ * unify this.
  */
 public class ServiceBusReceiveLinkProcessor extends FluxProcessor<AmqpReceiveLink, Message> implements Subscription {
-    // We don't want to dump too many credits on the link at once. It's easy enough to ask for more.
-    private static final int MINIMUM_REQUEST = 0;
-    private static final int MAXIMUM_REQUEST = 100;
-
     private final ClientLogger logger = new ClientLogger(ServiceBusReceiveLinkProcessor.class);
     private final Object lock = new Object();
     private final AtomicBoolean isTerminated = new AtomicBoolean();
-    private final AtomicBoolean hasDownstream = new AtomicBoolean();
     private final AtomicInteger retryAttempts = new AtomicInteger();
-    private final AtomicBoolean isRequested = new AtomicBoolean();
-    private final AtomicInteger linkCreditRequest = new AtomicInteger(1);
+    private final Deque<Message> messageQueue = new ConcurrentLinkedDeque<>();
+    private final AtomicBoolean hasFirstLink = new AtomicBoolean();
+    private final AtomicBoolean linkCreditsAdded = new AtomicBoolean();
 
     private final int prefetch;
     private final AmqpRetryPolicy retryPolicy;
-    private Disposable parentConnection;
+    private final Disposable parentConnection;
     private final AmqpErrorContext errorContext;
 
-    private volatile Subscription upstream;
-    private volatile CoreSubscriber<? super Message> downstream;
-
     private volatile Throwable lastError;
+    private volatile boolean isCancelled;
     private volatile AmqpReceiveLink currentLink;
     private volatile Disposable currentLinkSubscriptions;
     private volatile Disposable retrySubscription;
 
+    private volatile CoreSubscriber<? super Message> downstream;
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<ServiceBusReceiveLinkProcessor, CoreSubscriber> DOWNSTREAM =
+        AtomicReferenceFieldUpdater.newUpdater(ServiceBusReceiveLinkProcessor.class, CoreSubscriber.class,
+            "downstream");
```

Comment:
Same here - prefer to use `AtomicReference`.

Reply:
For SB, I'm not sure, because they create a separate receive link for each session and non-session receiver. But I guess at any given time, they'll have (maxConcurrentSessions). 🤔